File: app/src/main/java/com/kksionek/gdzietentramwaj/Repository/TramRepository.java
package com.kksionek.gdzietentramwaj.Repository;

import android.arch.lifecycle.LiveData;

import com.kksionek.gdzietentramwaj.dataSource.Room.FavoriteTram;
import com.kksionek.gdzietentramwaj.dataSource.Room.TramDao;
import com.kksionek.gdzietentramwaj.dataSource.TramDataWrapper;
import com.kksionek.gdzietentramwaj.dataSource.TramInterface;

import java.util.List;

import javax.inject.Inject;

public class TramRepository {
    private static final String TAG = "TramRepository";

    private final TramLiveData mTramLiveData;
    private final TramDao mTramDao;
    private final FavoriteLinesConsumer mFavoriteRepositoryAdder;

    @Inject
    public TramRepository(TramDao tramDao, TramInterface tramInterface) {
        mTramDao = tramDao;
        mFavoriteRepositoryAdder = new FavoriteLinesConsumer(mTramDao);
        mTramLiveData = new TramLiveData(tramInterface, mFavoriteRepositoryAdder);
    }

    public LiveData<TramDataWrapper> getDataStream() {
        return mTramLiveData;
    }

    public void forceReload() {
        mTramLiveData.forceReload();
    }

    public LiveData<List<FavoriteTram>> getAllFavTrams() {
        return mTramDao.getAllFavTrams();
    }

    public LiveData<List<String>> getFavoriteTrams() {
        return mTramDao.getFavoriteTrams();
    }

    public void setTramFavorite(String lineId, boolean favorite) {
        mTramDao.setFavorite(lineId, favorite);
    }
}

File: app/src/main/java/com/kksionek/gdzietentramwaj/repository/TramRepository.kt
package com.kksionek.gdzietentramwaj.repository

import com.kksionek.gdzietentramwaj.dataSource.NetworkOperationResult
import com.kksionek.gdzietentramwaj.dataSource.TramData
import com.kksionek.gdzietentramwaj.dataSource.TramInterface
import com.kksionek.gdzietentramwaj.dataSource.room.FavoriteTram
import com.kksionek.gdzietentramwaj.dataSource.room.TramDao
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.PublishSubject
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class TramRepository @Inject constructor(
    private val mTramDao: TramDao,
    private val tramInterface: TramInterface,
    private val favoriteRepositoryAdder: FavoriteLinesConsumer
) {
    private val dateFormat = SimpleDateFormat(
        "yyyy-MM-dd HH:mm:ss",
        Locale.getDefault()
    )
    private val dataTrigger: PublishSubject<Unit> = PublishSubject.create()

    val dataStream: Flowable<NetworkOperationResult<List<TramData>>> =
        dataTrigger.toFlowable(BackpressureStrategy.DROP)
            .startWith(Unit)
            .switchMap {
                Flowable.interval(0, 10, TimeUnit.SECONDS)
                    .subscribeOn(Schedulers.io())
                    .flatMap {
                        val cal = Calendar.getInstance()
                        cal.add(Calendar.MINUTE, -2)
                        val refDate = dateFormat.format(cal.time)
                        Observable.mergeDelayError(
                            tramInterface.trams()
                                .flatMapObservable { (list) -> Observable.fromIterable(list) },
                            tramInterface.buses()
                                .flatMapObservable { (list) -> Observable.fromIterable(list) }
                        )
                            .subscribeOn(Schedulers.io())
                            .filter { (time) -> refDate <= time }
                            .toList()
                            .doOnSuccess(favoriteRepositoryAdder)
                            .toNetworkOperationResult()
                            .toFlowable()
                    }
            }

    private fun <T> Single<T>.toNetworkOperationResult() =
        map { data -> NetworkOperationResult.Success(data) as NetworkOperationResult<T> }
            .onErrorReturn { NetworkOperationResult.Error(it) }

    val allFavTrams: Flowable<List<FavoriteTram>>
        get() = mTramDao.getAllFavTrams()

    val favoriteTrams: Flowable<List<String>>
        get() = mTramDao.getFavoriteTrams()

    fun forceReload() {
        dataTrigger.onNext(Unit)
    }

    fun setTramFavorite(lineId: String, favorite: Boolean) {
        mTramDao.setFavorite(lineId, favorite)
    }
}
